egs/wsj/s5/steps/dur_model/train.sh

```diff
@@ -0,0 +1,156 @@
+mkdir -p $dir/log
+
+if [ $stage -le 0 ]; then
+  echo "$0: Initializing the duration model and nnet..."
+
+  transmodel=$alidir/final.mdl
+  for f in $transmodel $phones_dir/roots.int $phones_dir/extra_questions.int; do
+    [ ! -f $f ] && echo "$0: Required file for initializing not found: $f" && exit 1;
+  done
+
+  if [ $max_duration == 0 ]; then
+    echo "$0: Determining max-duration..."
+    if [ ! -f $dir/all.ali ]; then
+      gunzip -c $alidir/ali.*.gz >> $dir/all.ali || exit 1;
+    fi
+
+    max_duration=`ali-to-phones --write-lengths $transmodel ark:$dir/all.ali \
+                  ark,t:- | awk -v min_count=$min_repeat_count -F';' \
+                  '{ for(i=1; i<=NF; i++){ \
+                       n=split($(i), a, " ");\
+                       duration=a[n];\
+                       counts[duration]++;\
+                       if(duration>max_duration) max_duration=duration; \
+                   }} END { for(d=3; d<=max_duration; d++) \
+                              if(counts[d]<min_count){\
+                                print d-1; exit 0;\
+                              }\
+                            print max_duration;\
+                            exit 0;}'` || exit 1;
+
+  fi
+  echo "$0: Max duration: $max_duration"
+
+  $cmd $dir/log/durmod_init.log \
+       durmodel-init --max-duration=$max_duration \
+                     --left-context=$left_context \
+                     --right-context=$right_context \
+                     $phones_dir/roots.int $phones_dir/extra_questions.int \
+                     $durmodel || exit 1;
+
+fi
+
+if [ $stage -le 1 ]; then
+  echo "$0: Preparing examples..."
+  steps/dur_model/get_examples.sh --cmd $cmd $egs_opts $alidir $dir || exit 1;
+fi
+
+
+if [ $stage -le 2 ]; then
+
+  [ ! -f $durmodel ] && echo "$0: Duration model file not found (have you completed stage 0?): $durmodel" && exit 1;
+  [ ! -f $dir/train.egs ] && echo "$0: Train examples file not found (have you completed stage 1?): $dir/train.egs" && exit 1;
+  [ ! -f $dir/val.egs ] && echo "$0: Validation examples file not found (have you completed stage 1?): $dir/val.egs" && exit 1;
+
+  if $use_gpu; then
+    train_queue_opt="--gpu 1"
+    if [ $cmd != "queue.pl" ]; then
+      echo "$0: WARNING: you are running on GPU but you are not using queue.pl."
+      echo -n "Are you sure you are not in a cluster? (if yes press ENTER, else Ctrl-c)"
+      read OK
+    fi
+    if ! cuda-compiled; then
+      echo "$0: WARNING: you are running with one thread but you have not compiled"
+      echo "   for CUDA.  You may be running a setup optimized for GPUs.  If you have"
+      echo "   GPUs and have nvcc installed, go to src/ and do ./configure; make"
+      exit 1
+    fi
+  else
+    parallel_train_opts="--use-gpu=no"
+  fi
+
+  # TODO(hhadian): config is appiled here. Should it be moved to stage 0?
+  # At stage 0, the user does not know the feat dim and ouput dim.
+  if [[ ! -z $nnet_config ]]; then
```

Comment:
BTW, if each epoch on a GPU takes less than a minute, I think it would be a
good idea to use the trick I discussed with cat'ing the file many times, to
have each iteration corresponds to several epochs-- the number of actual
epochs could be determined by the script after measuring how much data you
have.  Aim to have each iteration take a couple of minutes.

Dan


On Sun, Jan 17, 2016 at 2:45 PM, Hossein Hadian <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/dur_model/train.sh
> <https://github.com/kaldi-asr/kaldi/pull/414#discussion_r49951083>:
>
> > +      echo -n "Are you sure you are not in a cluster? (if yes press ENTER, else Ctrl-c)"
> > +      read OK
> > +    fi
> > +    if ! cuda-compiled; then
> > +      echo "$0: WARNING: you are running with one thread but you have not compiled"
> > +      echo "   for CUDA.  You may be running a setup optimized for GPUs.  If you have"
> > +      echo "   GPUs and have nvcc installed, go to src/ and do ./configure; make"
> > +      exit 1
> > +    fi
> > +  else
> > +    parallel_train_opts="--use-gpu=no"
> > +  fi
> > +
> > +  # TODO(hhadian): config is appiled here. Should it be moved to stage 0?
> > +  # At stage 0, the user does not know the feat dim and ouput dim.
> > +  if [[ ! -z $nnet_config ]]; then
>
> OK. I will change it so that stage 0 corresponds with epoch 0.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/414/files#r49951083>.
>